Metadata queries with configurable server-side timeouts

wprzytula opened the pull request:
## Motivation
Some users tune their server-side timeouts so that they are tighter. For clusters with a large schema, however, this sometimes made schema queries time out. This PR excludes metadata queries from the default server-side timeout by overriding it with a custom one, to prevent timeouts upon querying metadata.

## What's done

**Adds `USING TIMEOUT` clause to metadata queries when applicable.**

By _metadata_ I mean schema + topology. I did it for both for consistency, even though only schema was required in the [issue](#1052)).

## Implementation

To ensure that no fetches are accidentally omitted from having the timeout added, both now and in the future, I created a new abstraction: `ControlConnection`, which is just a wrapper over `Arc<Connection>` that exposes some methods of `Arc<Connection>` (`query_iter`, `execute_iter`, `prepare`, and some minor getters), taking care of adding the timeout before execution if applicable.

A corresponding setting is added to `SessionConfig`, and `SessionBuilder` gets a new method for configuring it.

## Testing

I added two tests:
1. A unit test of `ControlConnection`, which asserts that for ScyllaDB the timeout is enforced (if set) and for Cassandra is always ignored, in the following cases:
    - when explicitly disabled (no custom timeout follows),
    - when explicitly set to some (only set for ScyllaDB).
2. An integration test for the driver, which asserts that for ScyllaDB the timeout is enforced (if set) and for Cassandra is always ignored, in the following cases:
    - when explicitly disabled (no custom timeout follows),
    - when explicitly set to some (only set for ScyllaDB),
    - when left as implicit default (only set to ScyllaDB).

Fixes: #1052

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
As you can see, our SSL and Authenticate workflows fail, because the ScyllaDB run there is so old that it does not recognize `USING TIMEOUT` syntax...:
- [scylladb/scylla-passauth](https://hub.docker.com/r/scylladb/scylla-passauth)
  •Updated almost 4 years ago
- [scylladb/scylla-tls](https://hub.docker.com/r/scylladb/scylla-tls)
  •Updated almost 4 years ago

I guess we **urgently** need to create new images, to really test that the driver is compatible with those features in the new Scyllas, not 4-years-old ones...

dkropachev commented:
> As you can see, our SSL and Authenticate workflows fail, because the ScyllaDB run there is so old that it does not recognize `USING TIMEOUT` syntax...:
> 
> * [scylladb/scylla-passauth](https://hub.docker.com/r/scylladb/scylla-passauth)
>   •Updated almost 4 years ago
> * [scylladb/scylla-tls](https://hub.docker.com/r/scylladb/scylla-tls)
>   •Updated almost 4 years ago
> 
> I guess we **urgently** need to create new images, to really test that the driver is compatible with those features in the new Scyllas, not 4-years-old ones...
> 
> cc @dkropachev

From what I see these are regular images, based of scylladb/scylla/4.3.rc0 with some patches at `/etc/scylla` on top of it.
What I recomend you doing, create a configuration that you need to have at `/etc/scylla/` put it in repo and mount it at `/etc/scylla`,  and use scylladb/scylla as an image.

dkropachev commented:
> > As you can see, our SSL and Authenticate workflows fail, because the ScyllaDB run there is so old that it does not recognize `USING TIMEOUT` syntax...:
> > 
> > * [scylladb/scylla-passauth](https://hub.docker.com/r/scylladb/scylla-passauth)
> >   •Updated almost 4 years ago
> > * [scylladb/scylla-tls](https://hub.docker.com/r/scylladb/scylla-tls)
> >   •Updated almost 4 years ago
> > 
> > I guess we **urgently** need to create new images, to really test that the driver is compatible with those features in the new Scyllas, not 4-years-old ones...
> > cc @dkropachev
> 
> From what I see these are regular images, based of scylladb/scylla/4.3.rc0 with some patches at `/etc/scylla` on top of it. What I recomend you doing, create a configuration that you need to have at `/etc/scylla/` put it in repo and mount it at `/etc/scylla`, and use scylladb/scylla as an image.

Done up [here](https://github.com/scylladb/scylla-rust-driver/pull/1173)

muzarski reviewed (approved):
LGTM. Nice piece of code - especially the test cases

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -2406,7 +2406,7 @@ mod tests {
     use std::time::Duration;
 
     // Just like resolve_hostname in session.rs
-    async fn resolve_hostname(hostname: &str) -> SocketAddr {
+    pub(crate) async fn resolve_hostname(hostname: &str) -> SocketAddr {
         match tokio::net::lookup_host(hostname).await {
             Ok(mut addrs) => addrs.next().unwrap(),
```
🔧 Instead of pub(crate)'ing it please move it to test_utils - this is what this module is for.

Lorak-mmk commented on scylla/src/transport/topology.rs:
```diff
@@ -869,6 +870,181 @@ mod control_connection {
+
+        use super::ControlConnection;
+
+        /// Tests that ControlConnection enforces the provided custom timeout
+        /// iff ScyllaDB is the target node (else ignores the custom timeout).
+        #[cfg(not(scylla_cloud_tests))]
+        #[tokio::test]
+        #[ntest::timeout(2000)]
+        async fn test_custom_timeouts() {
+            setup_tracing();
+
+            let proxy_addr = SocketAddr::new(scylla_proxy::get_exclusive_local_address(), 9042);
+            let uri = std::env::var("SCYLLA_URI").unwrap_or_else(|_| "127.0.0.1:9042".to_string());
+            let node_addr: SocketAddr = resolve_hostname(&uri).await;
+
```
❓ This test requires a running Scylla / C* cluster, right? Is it possible to avoid this?

wprzytula replied:
It is possible to avoid this by using the proxy in the _dry mode_. We could mock the whole CQL handshake (`OPTIONS -> SUPPORTED -> STARTUP -> READY -> REGISTER -> READY`) and then intercept `QUERY`s and `PREPARE`s as before. An additional gain would be to test multiple scenarios at ones, i.e. sharded and non-sharded endpoints, independently of the actual cluster being deployed for tests.

wprzytula replied:
I've made the test use the dry mode to avoid dependency on a running cluster.

Lorak-mmk commented on scylla/src/transport/topology.rs:
```diff
@@ -743,28 +743,30 @@ impl MetadataReader {
-) -> Result<Metadata, QueryError> {
-    let peers_query = query_peers(conn, connect_port);
-    let keyspaces_query = query_keyspaces(conn, keyspace_to_fetch, fetch_schema);
+impl ControlConnection {
+    async fn query_metadata(
+        self,
+        connect_port: u16,
+        keyspace_to_fetch: &[String],
+        fetch_schema: bool,
+    ) -> Result<Metadata, QueryError> {
+        let peers_query = self.clone().query_peers(connect_port);
+        let keyspaces_query = self.query_keyspaces(keyspace_to_fetch, fetch_schema);
```
Commit: "topology: make metadata fetchers methods on ControlConnection "

🔧 For me it is extremely weird and unintuitive to have multiple impl blocks for a struct (ControlConnection), scattered in various places of 2 modules (control_connection and topology).
Can we please put all the `impl ControlConnection` into the `control_connection` module, and preferably make them a single `impl` block?

wprzytula replied:
https://www.reddit.com/r/rust/comments/w5l320/is_having_multiple_impl_blocks_idiomatic/

It's a perfectly valid way of logical separation. In this particular case, we separate metadata-related functionalities (defined in the impl block in `metadata.rs` main module) from the lower-level functionalities (defined in the impl block in `control_connection` module). Also, what we get is that methods defined out of `control_connection` module **can't access** mod-private fields and methods of `ControlConnection`.

Lorak-mmk commented on scylla/src/transport/topology.rs:
```diff
@@ -743,28 +743,30 @@ impl MetadataReader {
-) -> Result<Metadata, QueryError> {
-    let peers_query = query_peers(conn, connect_port);
-    let keyspaces_query = query_keyspaces(conn, keyspace_to_fetch, fetch_schema);
+impl ControlConnection {
+    async fn query_metadata(
+        self,
+        connect_port: u16,
+        keyspace_to_fetch: &[String],
+        fetch_schema: bool,
+    ) -> Result<Metadata, QueryError> {
+        let peers_query = self.clone().query_peers(connect_port);
```
🌱 When porting the PR to the `main` branch, after applying my previous comment about ControlConnection impls, I think it would be a good idea to extract `control_connection` module to a separate file.

wprzytula replied:
I think `ControlConnection` could be extracted to a separate file, but with those higher-level methods (`query_peers` etc.) left in the `metadata.rs` file. The reason is that `ControlConnection` itself is a medium, which should be logically separate from the logic performing queries using it.

wprzytula commented:
v1.1: resolved comments,
- made the unit test not require a running cluster.

roydahan commented:
There is no point in merging this into 0.15. branch, we're not going to release it another 0.15 release.

Lorak-mmk commented on scylla/src/cluster/metadata.rs:
```diff
@@ -601,7 +601,8 @@ impl MetadataReader {
         self.control_connection.wait_until_initialized().await;
-        let conn = ControlConnection::new(self.control_connection.random_connection()?);
+        let conn = ControlConnection::new(self.control_connection.random_connection()?)
+            .override_timeout(None);
```
Commit "metadata: customise ControlConnection's request timeout"

Why call `.override_timeout(None)` here? It is none even without this call.

wprzytula replied:
It's there just for explicitness. The `None` serves as a placeholder, which is replaced by an actual value in the next commit.

Lorak-mmk commented on scylla/src/cluster/metadata.rs:
```diff
@@ -773,30 +777,62 @@ mod control_connection {
     pub(super) struct ControlConnection {
         conn: Arc<Connection>,
+        overriden_timeout: Option<Duration>,
     }
```
Could you change the name to indicate that it is about server-side timeout?

wprzytula replied:
Very good idea.

wprzytula replied:
Done in a new commit.

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -278,6 +281,7 @@ impl SessionConfig {
             timestamp_generator: None,
             keyspaces_to_fetch: Vec::new(),
             fetch_schema_metadata: true,
+            metadata_request_serverside_timeout: Some(Duration::from_secs(2)),
             keepalive_interval: Some(Duration::from_secs(30)),
             keepalive_timeout: Some(Duration::from_secs(30)),
```
I wonder if this is the right default value. As far as I understand, the goal is to prevent wasting work (which is wasted, if the cluster is overloaded and we hit the timeout).
Maybe 20 seconds would be better?

cc @mykaul

Lorak-mmk reviewed (changes requested):
It's awesome that you made the in-crate tests not require real db!

wprzytula commented on scylla/src/cluster/metadata.rs:
```diff
@@ -813,277 +814,284 @@ mod control_connection {
-            table: "system.peers",
-        })
-        .and_then(|row_result| future::ok((NodeInfoSource::Peer, row_result)));
+        let local_ip: IpAddr = self.get_connect_address().ip();
+        let local_address = SocketAddr::new(local_ip, connect_port);
```
`self` is consumed below the new location of `local_ip`. If I were to leave it where it had been before, I would have to clone the `Arc` once more, which I preferred to avoid.

wprzytula commented on scylla/src/cluster/metadata.rs:
```diff
@@ -710,28 +710,30 @@ impl MetadataReader {
-) -> Result<Metadata, MetadataError> {
-    let peers_query = query_peers(conn, connect_port);
-    let keyspaces_query = query_keyspaces(conn, keyspace_to_fetch, fetch_schema);
+impl ControlConnection {
+    async fn query_metadata(
+        self,
+        connect_port: u16,
+        keyspace_to_fetch: &[String],
```
> ❓ Why does it take `self`? Previously it took `&Arc<Connection>` and I see no mention of this change in commit message.

It takes `self`, because it's supposed to be a method on `ControlConnection`, so that `ControlConnection` can hide its implementation details (at this time: just applying the custom server-side timeout) when executing metadata queries.

> Imo ControlConnection should not be possible to clone at all, hinting that we can't have multiple of themm

Although I understand your concern, I still believe the approach is justified. 
1. `ControlConnection` is clonable by sharing ownership only (because it has an `Arc` inside), so it's still keeping the single control connection.
2. There are more types that have such shared ownership in the driver, most notably `ExecutionProfile` and `ExecutionProfileHandle`. `ExecutionProfile` is shared when cloned, and immutable, which makes it very similar to current `ControlConnection`.
3. What could be a better alternative? Would you prefer to

Lorak-mmk replied:
> It takes self, because it's supposed to be a method on ControlConnection, so that ControlConnection can hide its implementation details (at this time: just applying the custom server-side timeout) when executing metadata queries.

What I meant is why it takes `self` instead of `&self`. Before the change it took a reference. I don't see why you decided to change this.

wprzytula replied:
Then the answer is simple: it takes `self` because its callees take `self`. It's the same as what we would do about `String`: if a callee of `foo` takes `String`, then `foo` would take `String` too, not `&str` or `&String`, because this way we can avoid a clone.

muzarski replied:
> Then the answer is simple: it takes `self` because its callees take `self`. It's the same as what we would do about `String`: if a callee of `foo` takes `String`, then `foo` would take `String` too, not `&str` or `&String`, because this way we can avoid a clone.

But do the callees need to take `self`? Why can't all `query_*` methods accept `&self`?

muzarski replied:
Ahh, I see that `Connection` methods accept `self: Arc<Self>` (before this PR). I assumed that they accept `&self`. In such case, I think it's fine that `ControlConnectionMethods` accept `self` instead of `&self`.

wprzytula replied:
> Ahh, I see that `Connection` methods accept `self: Arc<Self>` (before this PR). I assumed that they accept `&self`. In such case, I think it's fine that `ControlConnection` methods accept `self` instead of `&self`.

Yep.

> If they accepted `&self`, we would need to `Arc::clone()` the underlying connection anyway. Maybe it's better? Since the `Arc::clone()` is explicit, and not hidden behind `derive(Clone)` of `ControlConnection`?

I believe the current behaviour is in line with the general rule of passing ownership to the function if only it needs to consume the object, which is the case here. That's why I'm still convinced to leave it in the current state.

wprzytula commented:
Removed `#[derive(Clone)]` from `ControlConnection` according to @Lorak-mmk's suggestion.
Rewrote commits so that there are no longer fixes in later commits.

muzarski reviewed (approved):
I see that in the end, we decided to accept `&self` in `ControlConnection` methods and `Arc::clone()` the underlying `Connection` internally. Makes sense.
